### DISCUSSION TITLE
Solving Tangent linear System

### DISCUSSION THREAD

**Discussion by nikhilgv91** [idaholab/moose#16934] [Q&A General] [answered]
Hello,

My goal is to perform PDE-constrained optimization, and I am using MOOSE to perform the simulations. In addition to the simulation result, I would like to compute the solution of the 'Tangent Linear System' using MOOSE. A short explanation for what I would like to do:
My PDE: F (T,v) = 0
T : state variable
v : control variable (real number)
Tangent linear system: [dF(T,v) / dT] * x = - [dF(T,v) / dv] ; linear solve for x

The matrix dF(T,v) / dT is the jacobian of the original PDE. I would like to know if I can modify the executioner (or other appropriate object) so that the tangent linear system can be solved right after the "original" MOOSE solve in a single run of the simulation. 

I am aware that MOOSE can perform matrix-free solves, but for my problem I am using a NEWTON solve.

Thanks in advance!

Kind regards,
Nikhil

**Comment by YaqiWang** [comment, marked as answer]
Yes. You can modify executioners in whatever way. Once you have [dF(T,v) / dT] and [dF(T,v) / dv], you can directly use PETSc to solve for x. If dF(T,v) / dT is fixed in this solve, you do not have to use MOOSE's PJFNK machinery which could be messy with your "original" system. I feel you can evaluate dF(T,v) / dv pretty easily with MOOSE residual evaluation and `Control` for `v`. Just my suggestions.

**Reply by nikhilgv91** [reply]
Thanks for  your reply!

I went through to code to try and figure out how the MOOSE performs the solves. I realized that the snes solve routines are called by the petsc_nonlinear_solver object in libmesh. I am not sure where (or if) the jacobian is constructed. Could you please give me some hint?

**Reply by YaqiWang** [reply]
The jacobian or matrix is stored here: `SparseMatrix<Number> * FEProblem::getNonlinearSystem()::sys()::matrix`. It is probably constructed by SNES wrapped in the petsc_nonlinear_solver during solve. The Jacobian or matrix in the nonlinear implicit system is a `PetscMatrix<Real>`  which has a member function `mat()` to return a raw PETSc matrix pointer in type of `Mat`. But MOOSE allows you to construct a matrix by yourself with the function `Moose::compute_jacobian` in `NonlinearSystem.C`. This is my understanding.

**Reply by nikhilgv91** [reply]
I was able to access the jacobian as a SparseMatrix<Number> object. You mentioned in your earlier message that a vector can be computed using MOOSE residual evaluation. Could you please give more info about that? I see that there is a computeResidual function in NonlinearSystem class. Can this function be triggered through a kernel with a specific tag defined in the input file?

**Reply by aeslaughter** [reply]
There is a tagging system: https://mooseframework.inl.gov/framework_development/tagging.html

**Reply by nikhilgv91** [reply]
Thanks for this hint. Are there more detailed tutorials on the use of tagging? I looked at some input files in the module tests that uses tagging, but I don't understand it very well.

**Reply by YaqiWang** [reply]
I was thinking to use two residual evaluations to evaluate that one with v another with v+epsilon. Then take the difference of the two divided by epsilon.

**Reply by nikhilgv91** [reply]
The analytical derivative is trivial to compute, so I would prefer to assemble the residual using the analytical expression.

**Reply by nikhilgv91** [reply]
@aeslaughter and @YaqiWang thanks for your respective suggestions! I was able to solve the Tangent Linear System. I will mark this thread as the solution to my query.

**Comment by aeslaughter** [comment]
@zachmprince @lynnmunday How does this relate to your optimization work?

**Reply by zachmprince** [reply]
This is somewhat related. @lynnmunday could speak more to it. Basically we are approximating  [dF(T,v) / dv]  using an adjoint equation evaluation, at least that is my understanding.

**Reply by nikhilgv91** [reply]
Is PDE-constrained optimization already already part of the modules?

**Reply by lynnmunday** [reply]
We have an app that is doing pde constrained optimization but it needs cleaned up with better testing before we can merge it into moose.  I think we would have it merged in by the end of March.  We are use TAO to do the optimization and we are providing it with a forward and gradient solve.  The gradient solve is computed from the adjoint which we get by just solving the system a second time with the adjoint loads and bcs.  We do this with the multiapp system.

**Reply by nikhilgv91** [reply]
@lynnmunday Thanks for the detailed reply!
